### ISSUE TITLE
sog parser is a turd

### ISSUE THREAD

**Issue by RemoveRusky** [enigma-dev/enigma-dev#1187]
sog parser breaks on simple for loop

for (int i = 0; i < 4; i++) {
}

Check `object0::create...Syntax error in object `object0', Create event:0:
Line 1, position 27 (absolute 26): Expected secondary expression before closing parenthesis

**Comment by JoshDreamland** [comment]
Recall that `++` is not a GML construct. You will have to enable it in settings. I agree that SOG should use EDL settings by default.

**Comment by RemoveRusky** [comment]
Also,

string str = "wat";
game_end();

 compiles but game_end() is never called.

**Comment by JoshDreamland** [comment]
What is the codegen emitted?

**Comment by RemoveRusky** [comment]
which file is that in?

**Comment by RemoveRusky** [comment]
apparently its not generating the code in the .edl

/********************************************************************************\
**                                                                              **
**  Copyright (C) 2008 Josh Ventura                                             **
**                                                                              **
**  This file is a part of the ENIGMA Development Environment.                  **
**                                                                              **
**                                                                              **
**  ENIGMA is free software: you can redistribute it and/or modify it under the **
**  terms of the GNU General Public License as published by the Free Software   **
**  Foundation, version 3 of the license or any later version.                  **
**                                                                              **
**  This application and its source code is distributed AS-IS, WITHOUT ANY      **
**  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS   **
**  FOR A PARTICULAR PURPOSE. See the GNU General Public License for more       **
**  details.                                                                    **
**                                                                              **
**  You should have recieved a copy of the GNU General Public License along     **
**  with this code. If not, see <http://www.gnu.org/licenses/>                  **
**                                                                              **
**  ENIGMA is an environment designed to create games and other programs with a **
**  high-level, fully compilable language. Developers of ENIGMA or anything     **
**  associated with ENIGMA are in no way responsible for its users or           **
**  applications created by its users, or damages caused by the environment     **
**  or programs made in the environment.                                        **
**                                                                              **
\********************************************************************************/

//This file was generated by the ENIGMA Development Environment.
//Editing it is a sign of a certain medical condition. We're not sure which one.


#define log_xor || log_xor_helper() ||
struct log_xor_helper { bool value; };
template<typename LEFT> log_xor_helper operator ||(const LEFT &left, const log_xor_helper &xorh) { log_xor_helper nxor; nxor.value = (bool)left; return nxor; }
template<typename RIGHT> bool operator ||(const log_xor_helper &xorh, const RIGHT &right) { return xorh.value ^ (bool)right; }

variant enigma::OBJ_object0::myevent_create()
{
  enigma::debug_scope $current_scope("event 'create' for object 'object0'");
enigma::temp_event_scope ENIGMA_PUSH_ITERATOR_AND_VALIDATE(this);
  toString();
  
  return 0;
}
variant enigma::OBJ_object0::myevent_draw()
{
  enigma::debug_scope $current_scope("event 'draw' for object 'object0'");
;
  
  return 0;
}
inline bool enigma::OBJ_object0::myevent_draw_subcheck()
{
  return (visible);

}
bool enigma::OBJ_object0::can_cast(int obj) const {
  return false;
}

namespace enigma
{
  callable_script callable_scripts[] = {
  };
  
  void constructor(object_basic* instance_b) {
    //This is the universal create event code
    object_locals* instance = (object_locals*)instance_b;
    
    instance->xstart = instance->x;
    instance->ystart = instance->y;
    instance->xprevious = instance->x;
    instance->yprevious = instance->y;
    
    instance->gravity=0;
    instance->gravity_direction=270;
    instance->friction=0;
    
    
    instance->timeline_index = -1;
    instance->timeline_running = false;
    instance->timeline_speed = 1;
    instance->timeline_position = 0;
    instance->timeline_loop = false;
    
    instance->image_alpha = 1.0;
    instance->image_angle = 0;
    instance->image_blend = 0xFFFFFF;
    instance->image_index = 0;
    instance->image_speed  = 1;
    instance->image_xscale = 1;
    instance->image_yscale = 1;
    
    instancecount++;
    instance_count++;
  }
}

**Comment by JoshDreamland** [comment]
But only when you declare a string in it?

**Comment by RemoveRusky** [comment]
yep

**Comment by JoshDreamland** [comment]
This should also happen when built through LGM, then. Does it not?

**Comment by RemoveRusky** [comment]
yes, it seems it's also bugged in lgm

**Comment by RemoveRusky** [comment]
without the word string it will compile

**Comment by JoshDreamland** [comment]
What if you put code before the string declaration? Or more at the end?

**Comment by RemoveRusky** [comment]
```
game_end();
game_end();
string str = "wat";
game_end();
game_end();
```
still errors

**Comment by JoshDreamland** [comment]
Errors? Or is empty?

**Comment by RemoveRusky** [comment]
Ok, so behaviors lil wacky eveywhere.

lgm will print ```
Check `obj_0::create...Syntax error in object `obj_0', Create event:0:
Line 1, position 1 (absolute 0): Unterminated brace at this point
```and stop compile regardless of if theres code before or after. The cli will compile regardless and generates code before the string  but puts a single `toString()` with no params where the string should be written and writes nothing after that.

**Comment by JoshDreamland** [comment]
This is probably due to the fact that `string` is both a type and a function. We've had this problem before, but I thought it had been solved.

**Comment by RemoveRusky** [comment]
well, clearly we need a `string str = string(5); `test

**Comment by JoshDreamland** [comment]
Let's maybe wait until I fix it again, for that.

**Comment by RemoveRusky** [comment]
this might be related to that stl cleanup I did. I think i moved so toString crap into a namepace

**Comment by JoshDreamland** [comment]
As I recall, the fix was `#define string(x) toString(x)`. Unless you deleted that, it should still be working.

**Comment by RemoveRusky** [comment]
grepping reveals `#define string(VALUE_TO_MAKE_A_STRING_OUT_OF...) toString(VALUE_TO_MAKE_A_STRING_OUT_OF)` in var4.h

**Comment by JoshDreamland** [comment]
Yeah, that looks perfect. Something else is up.

**Comment by RobertBColton** [comment]
> Recall that ++ is not a GML construct.

It was added in GMS 1.4, and still exists in GMS2. So you can start considering it a GML construct now. It does not help us to have that unsupported, nobody wants to flip a switch to enable it anymore.

**Comment by RobertBColton** [comment]
Actually come to think of it, looks like I actually set that as the default in the settings.ey a while ago:
https://github.com/enigma-dev/enigma-dev/blob/master/settings.ey#L19

What is + for GML mean? You mean `x+` would actually increment x in older versions of GameMaker? I also used `x += 1` format in GM because I had never heard of such a thing. Is this true?

**Comment by RobertBColton** [comment]
I don't think it's true, I just tested in GM5 and I got invalid assignment operator with this:
```gml
x+

show_message(x);
```

**Comment by JoshDreamland** [comment]
No; that code adds the result of `show_message(x)` to `x`. You need an assignment before it, though, since GM disallows useless statements. Remember that `a = b++` is valid, too. I have seen games pad numbers using the `+` symbol, since unary plus has no effect. So in those games, `a=b++1` (for any number of pluses you like) will create a problem if you enable incrementation operators.

That said, I agree that we should enable these operators by default. I was just pointing out to cheeseboy that the default is "0," which means "no incrementation."

I also believe we should enable "test undefined variables as value zero" by default, since the special behavior is actually required in order to *not* do that.

**Comment by RobertBColton** [comment]
You are right, the following gives me `7` in both GM5 and GM8, but a syntax error in GMS 1.4, unless I remove the 2 which then results in `5` in GMS 1.4 and a syntax error in GM5 and GM8.

```
var b;
b = 5;
a = b++2;
show_message(string(a));
```
